### PR TITLE
Implement sync for credentials

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/488551667048375/1205082906974103/f 

### Description
Allows user to sync their credentials across devices.

### Steps to test this PR
see notes in https://app.asana.com/0/488551667048375/1205082906974103/f

### CODE REVIEW

**Comment by cmonfortep** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3366** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3366" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3366?utm_source=stack-comment).

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt [hunk 1]
```diff
@@ -87,6 +87,11 @@ interface AutofillStore {
      */
     suspend fun getAllCredentials(): Flow<List<LoginCredentials>>
 
+    /**
+     * Returns all credentials that have been modified since the given timestamp
+     */
+    suspend fun getAllCredentialsModifiedSince(modifiedSince: Long): List<LoginCredentials>
```
Comment (CDRussell) [inline autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: I'm assuming this `modifiedSince` is milliseconds since 1970, but would be good to make that more explicit here (e.g, add javadoc for the parameter)

Comment (CDRussell) [inline autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: Since changing a public API, should be following this right?: [[Doc] Android Public APIs](https://app.asana.com/0/1202561462274611/1205060262189411/f)

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt [hunk 2]
```diff
@@ -168,13 +172,21 @@ class SecureStoreBackedAutofillStore(
...
+    }
+
     override suspend fun getCredentialCount(): Flow<Int> {
         return secureStorage.websiteLoginDetailsWithCredentials().map { it.size }
     }
 
     override suspend fun deleteCredentials(id: Long): LoginCredentials? {
         val existingCredential = secureStorage.getWebsiteLoginDetailsWithCredentials(id)
         secureStorage.deleteWebsiteLoginDetailsWithCredentials(id)
+        credentialsSyncMetadata.onEntityRemoved(id)
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt:189]: Probably related to https://app.asana.com/0/0/1205110657412668/f

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncDataPersister.kt [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+        if (result is Success) {
+            pruneDeletedObjects(response.credentials.last_modified)
+        }
+
+        return result
+    }
+
+    private fun processEntries(
+        crendentials: CrendentialsSyncEntries,
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncDataPersister.kt:85]: typo: 
`crendentials` used in many places in this class instead of `credentials`

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+@SingleInstanceIn(AppScope::class)
+class CredentialsSyncMetadata @Inject constructor(
+    private val dao: CredentialsSyncMetadataDao,
+) {
+    fun addOrUpdate(entity: CredentialsSyncMetadataEntity) {
+        dao.insert(entity)
+    }
+
+    fun getSyncId(id: Long): String? {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:34]: it's nice where you have used `syncId` to make it explicit what kind of ID it is; being more explicit that the `id` is a `localId` might also be good in this class

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    fun getRemovedEntitiesSince(since: String): List<CredentialsSyncMetadataEntity> {
+        return dao.getRemovedIdsSince(since)
+    }
+
+    fun onEntityRemoved(id: Long) {
+        val syncId = dao.getSyncId(id)
+        if (syncId != null) {
+            syncId.deleted_at = DatabaseDateFormatter.iso8601()
+            dao.insert(syncId)
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:60]: is `insert` right here? maybe i'm missing the greater context, but seems confusing to call `insert` from a function dealing with deletions

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+class CredentialsLastModifiedWinsStrategy constructor(
+    private val credentialsSync: CredentialsSync,
+    private val credentialsSyncMapper: CredentialsSyncMapper,
+    private val dispatchers: DispatcherProvider,
+) : CredentialsMergeStrategy {
+    override fun processEntries(credentials: CrendentialsSyncEntries): SyncMergeResult<Boolean> {
+        Timber.d("Sync-autofill-Persist: ======= MERGING TIMESTAMP =======")
+        return kotlin.runCatching {
+            credentials.entries.forEach { entry ->
+                runBlocking(dispatchers.io()) {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:39]: do we have to use `runBlocking` here? which thread are we blocking?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import timber.log.Timber
+
+class CredentialsLastModifiedWinsStrategy constructor(
+    private val credentialsSync: CredentialsSync,
+    private val credentialsSyncMapper: CredentialsSyncMapper,
+    private val dispatchers: DispatcherProvider,
+) : CredentialsMergeStrategy {
+    override fun processEntries(credentials: CrendentialsSyncEntries): SyncMergeResult<Boolean> {
+        Timber.d("Sync-autofill-Persist: ======= MERGING TIMESTAMP =======")
+        return kotlin.runCatching {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:37]: checking my understanding here: I guess this means if there is any error with any entry, we call the whole merge operation a failure? i.e, we don't want to ignore a failing entry and process the rest?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import com.duckduckgo.sync.api.engine.SyncMergeResult.Success
+import kotlinx.coroutines.runBlocking
+import timber.log.Timber
+
+class CredentialsLastModifiedWinsStrategy constructor(
+    private val credentialsSync: CredentialsSync,
+    private val credentialsSyncMapper: CredentialsSyncMapper,
+    private val dispatchers: DispatcherProvider,
+) : CredentialsMergeStrategy {
+    override fun processEntries(credentials: CrendentialsSyncEntries): SyncMergeResult<Boolean> {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:35]: There's a lot going on in this method; finding it hard to process (pun intended) all that it's doing. Not sure if you could refactor out some logic into standalone functions perhaps to help break it down a bit.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ */
+
+package com.duckduckgo.autofill.sync.persister
+
+import com.duckduckgo.autofill.sync.CrendentialsSyncEntries
+import com.duckduckgo.sync.api.engine.SyncMergeResult
+
+interface CredentialsMergeStrategy {
+    fun processEntries(credentials: CrendentialsSyncEntries): SyncMergeResult<Boolean>
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt:23]: The API gives no indication that it will be expensive to run or block the calling thread internally. We should at least mark this with a worker thread annotation?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+import com.duckduckgo.sync.api.engine.SyncMergeResult.Success
+import kotlinx.coroutines.runBlocking
+import timber.log.Timber
+
+class CrendentialsDedupStrategy constructor(
+    private val credentialsSync: CredentialsSync,
+    private val credentialsSyncMapper: CredentialsSyncMapper,
+    private val dispatchers: DispatcherProvider,
+) : CredentialsMergeStrategy {
+    override fun processEntries(credentials: CrendentialsSyncEntries): SyncMergeResult<Boolean> {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt:36]: Same comment as on the other merge strategy; having everything in one large function is hurting readability of the code IMO

**Review by CDRussell — COMMENTED** [review]
Still working through this review but figure i'll flush out the pending comments so far for consideration

File: autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+interface CredentialsSyncMetadataDao {
+
+    @Query("select * from credentials_sync_meta where id = :id")
+    fun getSyncId(id: Long): CredentialsSyncMetadataEntity?
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    fun insert(entity: CredentialsSyncMetadataEntity): Long
+
+    @Query("select * from credentials_sync_meta where deleted_at >= :since")
+    fun getRemovedIdsSince(since: String): List<CredentialsSyncMetadataEntity>
```
Comment (CDRussell) [inline autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt:35]: we should mention the format of the string (presumably ISO-8601?)

File: secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt [hunk 12]
```diff
@@ -93,6 +93,19 @@ interface SecureStorage {
     @Throws(SecureStorageException::class)
     suspend fun websiteLoginDetailsWithCredentials(): Flow<List<WebsiteLoginDetailsWithCredentials>>
 
+    /**
+     * This method returns all the [WebsiteLoginDetailsWithCredentials] stored in the [SecureStorage] modified [since].
+     * This returns decrypted sensitive data (encrypted in L2). Use this only when sensitive data is needed to be accessed.
+     * If [canAccessSecureStorage] is false when this is invoked, an empty flow will be emitted.
```
Comment (CDRussell) [inline secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt:99]: > If [canAccessSecureStorage] is false when this is invoked, an empty flow will be emitted.

if this does happen, what will sync do with that?

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt [hunk 13]
```diff
@@ -18,6 +18,7 @@ package com.duckduckgo.sync.api.engine
 
 import com.duckduckgo.sync.api.engine.SyncableType.BOOKMARKS
 
+// TODO: document api
```
Comment (CDRussell) [inline sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt:21]: you should do it 😄

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/SyncApiClient.kt [hunk 14]
```diff
@@ -83,6 +84,10 @@ class AppSyncApiClient @Inject constructor(
 
         return when (type) {
             BOOKMARKS -> getBookmarks(type, token, since)
+            CREDENTIALS -> getCredentials(type, token, since)
+            else -> {
+                TODO()
```
Comment (CDRussell) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/SyncApiClient.kt:89]: kill this?

**Review by CDRussell — COMMENTED** [review]
.

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt (hunk 1 above)
Comment (cmonfortep) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: This is now removed as I'm moving away from using AutofillStore (as discussed)

Comment (cmonfortep) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: This is now removed as I'm moving away from using AutofillStore (as discussed)

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt (hunk 2 above)
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt:189]: all good here (when user deletes, we update delete timestamp in metadata db)

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt (hunk 4 above)
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:34]: also updated at db level, sounds good

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt (hunk 5 above)
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:60]: discussed offline: it's updating `deleted_at` timestamp

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt (hunk 6 above)
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:39]: annotated the interface with WorkerThread. Strategies are used as part of Persister, which also checks we are not in main thread.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt (hunk 9 above)
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt:23]: this is now annotated thanks.

File: secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt (hunk 12 above)
Comment (cmonfortep) [reply secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt:99]: Since we have timestamp in metadata table, we don't need to use this method anymore 🙌

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+        credentialsSyncStore.clientModifiedSince = "0"
+
+        secureStorage.websiteLoginDetailsWithCredentials().firstOrNull()?.map { it.details.id }?.filterNotNull()?.let { autofillIds ->
+            credentialsSyncMetadata.initializeDatabase(autofillIds)
+        }
+
+        Timber.i("CredentialsSync: initMetadata ${credentialsSyncMetadata.getAllObservable().firstOrNull()}")
+    }
+
+    suspend fun getUpdatesSince(since: String): List<LoginCredentialEntry> {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt:52]: Nit: this could be a good use case for [inline value classes](https://kotlinlang.org/docs/inline-classes.html)

the string parameter could accept any string, but has a very specific format required in order for it to be used. could define something like

`value class Iso8601String(private val s: String)` that is passed into these methods instead of just a string. might help make it clearer what is expected and avoid future errors by forcing the caller to be a little bit more deliberate in the string that is passed

just a suggestion to consider

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+    private suspend fun allContentAsUpdates() = secureStorage.websiteLoginDetailsWithCredentials().firstOrNull().mapToLoginCredentialEntry()
+
+    private suspend fun changesSince(since: String): List<LoginCredentialEntry> {
+        Timber.d("CredentialsSync: generating changes since $since")
+        credentialsSyncMetadata.getAllCredentials().forEach {
+            Timber.i("CredentialsSync: syncMetadata $it")
+        }
+
+        val values2 = credentialsSyncMetadata.getChangesSince(since).map {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt:138]: is there a better name than `values2`?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncStore.kt [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        set(value) = preferences.edit(true) { putString(KEY_START_TIMESTAMP, value) }
+    override var clientModifiedSince: String
+        get() = preferences.getString(KEY_END_TIMESTAMP, "0") ?: "0"
+        set(value) = preferences.edit(true) { putString(KEY_END_TIMESTAMP, value) }
+
+    private val preferences: SharedPreferences
+        get() = context.getSharedPreferences(FILENAME, Context.MODE_PRIVATE)
+
+    companion object {
+        const val FILENAME = "com.duckduckgo.credentials.sync.store"
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncStore.kt:50]: not a big deal, but should we match the file's name package to the filename here?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package com.duckduckgo.autofill.sync
+
+import org.threeten.bp.OffsetDateTime
+import org.threeten.bp.ZoneOffset
+import org.threeten.bp.format.DateTimeFormatter
+
+class SyncDateProvider {
+    companion object {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt:24]: nit: it might make things easier to test if this was a interface we injected in DI, and could replace it with a double during testing. we do that elsewhere, e.g.

```
interface TimeProvider {
    fun currentTimeMillis(): Long
}

@ContributesBinding(AppScope::class)
class SystemCurrentTimeProvider @Inject constructor() : TimeProvider {
    override fun currentTimeMillis(): Long = System.currentTimeMillis()
}
```

 feel free to ignore if testing what happens at different times isn't a problem

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+import com.duckduckgo.autofill.sync.CredentialsSyncMapper
+import com.duckduckgo.autofill.sync.credentialsSyncEntries
+import com.duckduckgo.autofill.sync.isDeleted
+import com.duckduckgo.sync.api.engine.SyncMergeResult
+import com.duckduckgo.sync.api.engine.SyncMergeResult.Error
+import com.duckduckgo.sync.api.engine.SyncMergeResult.Success
+import kotlinx.coroutines.runBlocking
+import timber.log.Timber
+
+class credentialsDedupStrategy constructor(
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt:31]: Typo: class name should start with a capital letter

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    private val credentialsSyncMapper: CredentialsSyncMapper,
+    private val dispatchers: DispatcherProvider,
+) : CredentialsMergeStrategy {
+    override fun processEntries(
+        credentials: credentialsSyncEntries,
+        clientModifiedSince: String,
+    ): SyncMergeResult<Boolean> {
+        Timber.d("Sync-autofill-Persist: ======= MERGING DEDUPLICATION =======")
+
+        return kotlin.runCatching {
```
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt:42]: As per similar comment elsewhere, this code is hard to follow as is and could benefit from being pulled out into several smaller functions I reckon.

File: autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt (hunk 11 above)
Comment (CDRussell) [reply autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt:48]: bump

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt [hunk 21]
```diff
@@ -33,16 +34,19 @@ data class SyncChangesRequest(val type: SyncableType, val jsonString: String, va
...
     }
 }
 
+// TODO: document api
 enum class SyncableType(val field: String) {
     BOOKMARKS("bookmarks"),
+    CREDENTIALS("credentials"),
 }
 
+// TODO: document api, when is it expected each case?
```
Comment (CDRussell) [inline sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt:49]: do it?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt [hunk 22]
```diff
@@ -111,13 +111,13 @@ class RealSyncEngine @Inject constructor(
         getChanges().forEach {
             if (it.isEmpty()) {
                 Timber.d("Sync-Feature: ${it.type} local data empty, nothing to send")
-                syncStateRepository.updateSyncState(SUCCESS)
-                return
+                // syncStateRepository.updateSyncState(SUCCESS)
```
Comment (CDRussell) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt:114]: delete this?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt [hunk 23]
```diff
@@ -163,13 +164,13 @@ class RealSyncEngine @Inject constructor(
         return when (val result = syncApiClient.patch(changes)) {
             is Error -> {
                 Timber.d("Sync-Feature: patch failed ${result.reason}")
-                syncStateRepository.updateSyncState(FAIL)
+                // syncStateRepository.updateSyncState(FAIL)
```
Comment (CDRussell) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt:167]: Won't comment them all, but these comments should be deleted / clear up somehow

File: sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt [hunk 24]
```diff
@@ -79,6 +80,8 @@ internal class SyncEngineTest {
     }
 
     @Test
+    @Ignore
+    // https://app.asana.com/0/1204842202586359/1205158805627400/f
```
Comment (CDRussell) [inline sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt:84]: private task

**Review by CDRussell — APPROVED** [review]
Approving as the various edge cases all seem resolved (🎉).

There's still a few comments from last time that aren't addressed, and a few more to consider. Approving assuming they will all be addressed (fix or comment why not) before merging.

Looking good!

File: sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt (hunk 24 above)
Comment (cmonfortep) [reply sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt:84]: Public now 👍

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt (hunk 21 above)
Comment (cmonfortep) [reply sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt:49]: That was for me to create a task in Asana, not to fix it as part of this PR.

Changed comment to refer to asana instead 👍

File: autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt (hunk 11 above)
Comment (cmonfortep) [reply autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt:48]: Documented that in the Entity class, also in the `CredentialsSyncMetadata` class. Dao should be used only internally.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt (hunk 16 above)
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt:138]: `values007` jk 

good catch, I can go for `changes` instead.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt (hunk 18 above)
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt:24]: Just introduced this wrapper for clarity when reading the code, I'm good not injecting this for now.
